Add trivia categories.

CloakerSmoker opened the pull request:
Since the trivia API supports getting questions for a specific category, I've changed the first parameter of the `trivia` command to a category option, which is handled in the converter `CategoryConverter` where it is checked if the parameter is in the global dict `TRIVIA_CATEGORIES`, which maps a (string) category to an (int) trivia API category ID.

`TRIVIA_CATEGORIES` is built in the method `Games.get_trivia_categories`, which is called by `Games.__init__()`, and makes a request to the trivia API's category list endpoint and then builds the map of names/IDs out of the result, and also creates some extra entries in the map for categories which have much more common names than the API uses.

`CategoryConverter` also checks if the parameter is a difficulty level, which allows for the trivia command to be used like originally, with just a single param being the difficulty. In this version, the trivia difficulty is always the last param, so it's the first without a category, and the second with a category. 

The trivia sub-command `trivia categories` simply enumerates the `TRIVIA_CATEGORIES` dict, and prints out all categories in a simple embed. 

The only other change I've made is escaping any markdown in any trivia converter classes, which stops user-provided markdown from showing up in the error messages.

CloakerSmoker committed:
Added a method to get the category list from the trivia API, and a category converter

CloakerSmoker committed:
Added trivia category param to the trivia command, and place it before the difficulty param

CloakerSmoker committed:
Fixed duplicate category entries, added a command to list all categories

CloakerSmoker committed:
Removed fstrings, allowed for the first param to trivia to be a difficulty OR a category

CloakerSmoker committed:
Added some comments, cleaned up the category list

CloakerSmoker committed:
Updated trivia command descriptions

CloakerSmoker committed:
Changed some comments, escaped markdown in errors

CloakerSmoker committed:
Changed DifficultyConverter to escape markdown in error messages

Run1e commented on cogs/games.py:
```diff
@@ -259,6 +320,22 @@ def _calculate_score(self, pot, time_spent):
 		)
 
+	@trivia.command()
+	async def categories(self, ctx):
```
add a bot perm check for embed_links, like done in .trivia stats below

Run1e commented on cogs/games.py:
```diff
@@ -108,13 +128,50 @@ def __init__(self, bot):
 		self.config = ConfigTable(bot, 'trivia', ('guild_id', 'user_id'))
 		self.playing = set()
 
+		asyncio.run_coroutine_threadsafe(self.get_trivia_categories(), self.bot.loop)
```
change to `self.bot.loop.create_task(...)`

Run1e commented on cogs/games.py:
```diff
@@ -259,6 +320,22 @@ def _calculate_score(self, pot, time_spent):
 			ctx.guild.id, ctx.author.id, answered_at, question_hash, result
 		)
 
+	@trivia.command()
```
add a bot perm check for embed_links, like done in .trivia stats below

CloakerSmoker replied:
Done. I thought that embed_links was just for putting links inside of embeds, so I figured I could safely leave it out, but the check is added now.

Run1e commented on cogs/games.py:
```diff
@@ -108,13 +128,50 @@ def __init__(self, bot):
+		asyncio.run_coroutine_threadsafe(self.get_trivia_categories(), self.bot.loop)
+
+	async def get_trivia_categories(self):
+		global TRIVIA_CATEGORIES
```
is this really necessary? shouldn't it exist in globals()

Run1e replied:
best to store the categories as a cog member, and in the converter do ctx.cog.whatever

CloakerSmoker replied:
Wow, I had no clue that `ctx.cog` was a thing, that's super useful. 
I only had `TRIVIA_CATEGORIES` global since the converter and cog both need it, I've changed it to be the cog member `trivia_categories` now that it doesn't have to be global.

Run1e commented on cogs/games.py:
```diff
@@ -259,6 +320,22 @@ def _calculate_score(self, pot, time_spent):
+		# Just dump the list of valid categories into an embed and send it
+		description = ''
+
+		for category_name in TRIVIA_CATEGORIES:
```
list comprehension would be nice here. would also remove the need of setting description to an empty string beforehand

Run1e replied:
probably not needed but a check for whether the categories have been fetched first would be nice.

CloakerSmoker replied:
Good point, I've moved the code for building the category list into `get_trivia_categories`, since the list only needs to be built once, and made a default for it that tells the user there was an error getting the category list.

Run1e commented on cogs/games.py:
```diff
@@ -108,13 +128,50 @@ def __init__(self, bot):
+	async def trivia(self, ctx, category: CategoryConverter = None, *, difficulty: DifficultyConverter = None):
+		'''Trivia time! Optionally specify a difficulty or category and difficulty as arguments. Valid difficulties are `easy`, `medium` and `hard`. Valid categories can be listed with `trivia categories`.'''
+
+		if (category is not None) and (type(category) != int):
```
not needed with better converter logic

Run1e commented on cogs/games.py:
```diff
@@ -108,13 +128,50 @@ def __init__(self, bot):
 	@commands.cooldown(rate=1, per=300.0, type=commands.BucketType.member)
-	async def trivia(self, ctx, *, difficulty: DifficultyConverter = None):
-		'''Trivia time! Optionally specify a difficulty as argument. Valid difficulties are `easy`, `medium` and `hard`.'''
+	async def trivia(self, ctx, category: CategoryConverter = None, *, difficulty: DifficultyConverter = None):
```
use union typing in the command signature.

since you run the difficulty converter if the category converter fails, make a union instead

see here for more info: https://discordpy.readthedocs.io/en/latest/ext/commands/commands.html#typing-union

CloakerSmoker replied:
I didn't use a union, since I noticed that `typing` also provides `Optional`, which I think fits much better since the category *is* an optional parameter, and should be treated as difficulty when the converter fails (which is pretty much exactly what I manually implemented with all the ugly converter logic)

Run1e commented on cogs/games.py:
```diff
@@ -85,10 +87,28 @@ class Difficulty(Enum):
+				if Difficulty[category_name.upper()]:
+					return category_name
+			except:
+				raise commands.CommandError('\'{}\' is not a valid trivia category, see the categories command.'.format(category_name))
```
what about only running the argument through clean_content if it fails and is presented to the user?

CloakerSmoker replied:
Good idea, I have both converters doing that now.

Run1e commented on cogs/games.py:
```diff
@@ -108,13 +128,50 @@ def __init__(self, bot):
+		for category in res['trivia_categories']:
+			name = category['name'].lower() # Name is sometimes something like `GeneralCategory: SpecificCategory`, like `Entertainment: Books`, so we just trim that out if it's there
+			colon_pos = name.find(':')
+			trimmed_name = name[0 if colon_pos == -1 else colon_pos + 2:] # Cut out both the colon, and the space after it
```
I dislike how you're dealing with string splitting here - I would prefer .split(..., num=1)

CloakerSmoker replied:
I wasn't a fan of it either, but for some reason I managed to blank on `.split`, it is using that now

Run1e commented on cogs/games.py:
```diff
@@ -126,6 +183,10 @@ def __init__(self, bot):
 			#type='boolean'
 		)
 
+		if type(category) == int:
```
assuming aiohttp handles it nicely, this parameter can probably just go directly into the dict at creation

CloakerSmoker replied:
Unfortunately not, aiohttp throws when a param type isn't a string or int.

Run1e reviewed (changes requested):
Nice, however it begs the questions whether this is a good idea to begin with. The reason the trivia command doing scoring is to check peoples trivia skills, and that would be nullified if you can simply select a category you're good at.

I haven't given it too much thought, but maybe not reward points if a category is given by the user?

CloakerSmoker committed:
Made requested changes

CloakerSmoker commented:
Yeah, you're right about people being able to abuse a single category they are good at, but if picking a category makes it so they don't get any points, I don't think categories would ever really be used.

Instead, I think that just adding a penalty for picking a category would be better, so I've added that as `CATEGORY_PENALTY = 3`, and modified `_calculate_score` to divide the final score by `CATEGORY_PENALTY` if a category was added to the question request.

I'm not 100% sure I like this solution either though, so I'm open to any other ideas you have. I just don't want to give 0 points for using a category.

Other than that, I've made your requested changes, and/or left a comment on the specific change to explain why I might have done something a bit differently.
